llpc/lower/llpcSpirvLowerRayTracing.cpp

```diff
@@ -109,66 +109,96 @@ static unsigned TraceParamsTySize[] = {
     9, // 16, HitTriangleVertexPositions
 };
 
-// ====================================================================================================================
-// Get payload idx for TraceRayKHR instruction.
-unsigned getTraceRayParamPayloadIdx(void) {
-  return TraceRayParam::Payload;
-}
-
 // =====================================================================================================================
 SpirvLowerRayTracing::SpirvLowerRayTracing() : SpirvLowerRayQuery(false) {
 }
 
 // =====================================================================================================================
-// Create function and link traceRays traceRaysinline function
+// Process a trace ray call by creating (or get if created) an implementation function and replace the call to it.
 //
-// @param func : The trace function to create
-// @param opcode : OpTraceRayKHR or OpTraceNV to process
-template <> void SpirvLowerRayTracing::createRayTracingFunc<OpTraceRayKHR>(Function *func, unsigned opcode) {
-  // Empty function
-  assert(func->size() == 0);
-  func->setName(func->getName() + m_module->getName());
-  auto rayTracingContext = static_cast<RayTracingContext *>(m_context->getPipelineContext());
-  bool indirect = rayTracingContext->getIndirectStageMask() & ShaderStageComputeBit;
+// @param inst : The original call instruction
+void SpirvLowerRayTracing::processTraceRayCall(BaseTraceRayOp *inst) {
+  std::string mangledName(inst->getCalledFunction()->getName());
+  mangledName += ".impl";
 
-  if (opcode == OpTraceRayKHR) {
+  Value *accelStruct = inst->getAccelStruct();
+  Value *rayFlags = inst->getRayFlags();
+  Value *cullMask = inst->getInstanceInclusionMask();
+  Value *sbtOffset = inst->getRayContributionToHitGroupIndex();
+  Value *sbtStride = inst->getMultiplierForGeometryContribution();
+  Value *missIndex = inst->getMissShaderIndex();
+  Value *rayOrigin = inst->getOrigin();
+  Value *rayTMin = inst->getTMin();
+  Value *rayDir = inst->getDirection();
+  Value *rayTMax = inst->getTMax();
+  Value *payload = inst->getPayload();
+  Value *paq = inst->getPaq();
+
+  m_builder->SetInsertPoint(inst);
+
+  SmallVector<Value *, 12> implCallArgs({accelStruct, rayFlags, cullMask, sbtOffset, sbtStride, missIndex, rayOrigin,
+                                         rayTMin, rayDir, rayTMax, payload, paq});
+
+  auto newCall = m_builder->CreateNamedCall(mangledName, inst->getFunctionType()->getReturnType(), implCallArgs,
+                                            {Attribute::NoUnwind, Attribute::AlwaysInline});
+
+  inst->replaceAllUsesWith(newCall);
+
+  auto func = m_module->getFunction(mangledName);
+
+  if (func->isDeclaration()) {
+    enum : unsigned {
```

Comment:
I don't quite like the anonymous enum. Is this something we do regularly? If not, I'd prefer to at least give it a name.

Reply:
It seems not, getting the old ones back. (Can't recall why I decided to remove them at the first place :p)